File: source/client/web/invoice-book-web/public_html/js/aatl-ib/service/client-location-service.js
/* 
 * Copyright (c) 2019 Absolute Apogee Technologies Ltd. All rights reserved.
 * 
 * =============================================================================
 * Revision History:
 * Date         Author          Detail
 * -----------  --------------  ------------------------------------------------
 * 2019-Jan-07  GShokar         Created
 * =============================================================================
 */

"use strict";

$aatl_ib.ClientLocationService = {

    find: function (criteria, callback) {

        $aatl_ib.ApiService.get("client/location/find",
                criteria,
                function (res, err) {
                    if (err) {
                        callback(null, {messages: ["Faild to get client location search results: " + err]});

                    } else if (res.status === "failure") {
                        callback(null, {messages: [res.message]});
                    } else if (res.status === "success") {
                        let clients = JSON.parse(res.data);
                        callback(clients);
                    } else {
                        callback(null, {messages: ["Invalid response from server"]});
                    }
                });
    },

    get: function (number, callback) {

        if (number === undefined || number === null) {
            callback({
                number: "",
                name: "",
                address: {
                    address1: "",
                    address2: "",
                    city: "",
                    province: $aatl_ib.LookupService.defaultProvince.code,
                    postalCode: ""
                },
                contact: {
                    phone: "",
                    email: ""
                }
            });
        } else {
            $aatl_ib.ApiService.get("client/location/" + number,
                    {},
                    function (res, err) {
                        if (err) {
                            callback(null, {messages: ["Faild to get client location detail: " + err]});

                        } else if (res.status === "failure") {
                            callback(null, {messages: [res.message]});
                        } else if (res.status === "success") {
                            let client = JSON.parse(res.data);
                            callback(client);
                        } else {
                            callback(null, {messages: ["Invalid response from server"]});
                        }
                    });
        }
    },

    save: function (location, callback) {

        let err = $aatl_ib.ClientLocationService.validate(location);

        if (err.messages.length === 0) {

            $aatl_ib.ApiService.put("client/location",
                    location,
                    function (res, serverErr) {
                        if (serverErr) {

                            err.messages.push("Failed to save the client location: " + serverErr);
                            callback(location, err);

                        } else if (res.status === "failure") {

                            err.messages.push(res.message);
                            callback(location, err);

                        } else if (res.status === "success") {

                            if (Array.isArray(res.warningMessages) && res.warningMessages.length > 0) {
                                err.messages = res.warningMessages;
                            }

                            callback(JSON.parse(res.data), err);
                        } else {
                            callback(location, "Invalid response from server");
                        }
                    });

            return;
        }

        callback(location, err);
    },
    list: function (number, callback) {

        $aatl_ib.ApiService.get("client/location/list/" + number,
                {},
                function (res, err) {
                    if (err) {
                        callback(null, {messages: ["Faild to get client locations: " + err]});

                    } else if (res.status === "failure") {
                        callback(null, {messages: [res.message]});
                    } else if (res.status === "success") {
                        let client = JSON.parse(res.data);
                        callback(client);
                    } else {
                        callback(null, {messages: ["Invalid response from server"]});
                    }
                });

    },

    validate: function (location) {

        let err = {messages: []};

        if (location) {

            if ($aatl_ib.utils.isStringEmpty(location.name)) {
                err.messages.push("Please enter location name.");
            }

            if (location.address && $aatl_ib.utils.isStringEmpty(location.address.address1)) {
                err.messages.push("Please enter address line 1.");
            }

            if (location.address && $aatl_ib.utils.isStringEmpty(location.address.city)) {
                err.messages.push("Please enter city.");
            }

            if (location.address && $aatl_ib.utils.isStringEmpty(location.address.province)) {
                err.messages.push("Please enter province.");
            }

            if (location.address && $aatl_ib.utils.isStringEmpty(location.address.postalCode)) {
                err.messages.push("Please enter postal code.");
            }
        }

        return err;
    }
};

File: source/bl/invoice-book-service/src/main/java/ca/aatl/app/invoicebook/bl/ejb/ClientLocationService.java
/*
 * Copyright (c) 2019 Absolute Apogee Technologies Ltd. All rights reserved.
 * 
 * =============================================================================
 * Revision History:
 * Date         Author          Detail
 * -----------  --------------  ------------------------------------------------
 * 2019-Jan-09  GShokar         Created
 * =============================================================================
 */
package ca.aatl.app.invoicebook.bl.ejb;

import ca.aatl.app.invoicebook.data.jpa.dao.ClientLocationDao;
import ca.aatl.app.invoicebook.data.jpa.entity.Address;
import ca.aatl.app.invoicebook.data.jpa.entity.ClientLocation;
import ca.aatl.app.invoicebook.data.jpa.entity.Contact;
import ca.aatl.app.invoicebook.exception.DataValidationException;
import ca.aatl.app.invoicebook.util.AppUtils;
import java.util.List;
import javax.ejb.EJB;
import javax.ejb.LocalBean;
import javax.ejb.Stateless;

/**
 *
 * @author GShokar
 */

@Stateless
@LocalBean
public class ClientLocationService {

    @EJB
    ClientLocationDao dao;
    
    @EJB
    LookupService lookupService;
    
    @EJB
    SequenceService sequenceService;
            
    public List<ClientLocation> list(String clientNumber) {
        
        return dao.list(clientNumber);
    }

    public ClientLocation newEntity()  throws Exception{
        
        ClientLocation location = new ClientLocation();

        location.setActive(true);

        location.setAddress(new Address());
        
        location.getAddress().setCountry(lookupService.defaultCountry());
        location.getAddress().setProvince(lookupService.defaultProvince());

        location.setContact(new Contact());

        return location;
    }

    public void validate(ClientLocation entity) throws DataValidationException {
        boolean rtnValue = true;

        StringBuilder sb = new StringBuilder();

        if (entity != null) {

            if (AppUtils.isNullOrEmpty(entity.getName())) {
                rtnValue = false;
                sb.append(System.lineSeparator());
                sb.append("Enter the location name.");
            } else if (dao.isExists(entity.getId(), entity.getNumber(), entity.getName(), entity.getClient().getId())) {
                rtnValue = false;
                sb.append(System.lineSeparator());
                sb.append("This location name already exists.");
            }

            if (entity.getAddress() != null) {

                Address address = entity.getAddress();

                if (AppUtils.isNullOrEmpty(address.getAddress1())) {
                    rtnValue = false;
                    sb.append(System.lineSeparator());
                    sb.append("Enter the address.");
                }

                if (AppUtils.isNullOrEmpty(address.getCity())) {
                    rtnValue = false;
                    sb.append(System.lineSeparator());
                    sb.append("Enter the city.");
                }

                if (AppUtils.isNullOrEmpty(address.getPostalCode())) {
                    rtnValue = false;
                    sb.append(System.lineSeparator());
                    sb.append("Enter the postal code.");
                }
            }

            if (entity.getContact() != null) {
                Contact contact = entity.getContact();

                if (AppUtils.isNullOrEmpty(contact.getPhone())) {
                    rtnValue = false;
                    sb.append(System.lineSeparator());
                    sb.append("Enter the phone number.");
                }

            }
        }

        if (!rtnValue) {
            throw new DataValidationException(sb.toString());
        }
    }

    public ClientLocation find(String number) throws Exception {
        return dao.find(number);
    }

    public void save(ClientLocation clientLocation) throws Exception{
        beforeSave(clientLocation);

        dao.save(clientLocation);
    }

    private void beforeSave(ClientLocation entity) throws DataValidationException{
        validate(entity);

        if (entity.getId() == null) {
            entity.setAddedBy(entity.getLastUpdatedBy());
            entity.setAddedDate(entity.getLastUpdatedDate());
            entity.getGuid();

            synchronized (this) {
                entity.setNumber(sequenceService.next("ClientLocationNumber"));
            }
        }

        if (entity.getAddress() != null && entity.getAddress().getId() == null) {

            entity.getAddress().setAddedBy(entity.getLastUpdatedBy());
            entity.getAddress().setAddedDate(entity.getLastUpdatedDate());
            entity.getAddress().setLastUpdatedBy(entity.getLastUpdatedBy());
            entity.getAddress().setLastUpdatedDate(entity.getLastUpdatedDate());
            entity.getAddress().getGuid();
        }

        if (entity.getContact() != null && entity.getContact().getId() == null) {
            entity.getContact().setAddedBy(entity.getLastUpdatedBy());
            entity.getContact().setAddedDate(entity.getLastUpdatedDate());
            entity.getContact().setLastUpdatedBy(entity.getLastUpdatedBy());
            entity.getContact().setLastUpdatedDate(entity.getLastUpdatedDate());
            entity.getContact().getGuid();
        }
    }
    
}
